### PR TITLE
Simulate a reset of filters if no values are different from the default ones

### LINKED ISSUE #7250: Filter are not submitted when they are all equal to the defaults values.
I found one use-case not fixed by https://github.com/sonata-project/SonataAdminBundle/pull/6871

Let's say you have two filter A and B.
Let's say the default situation is "The list is filtered by filter A with the value "TRUE"".

If I change the filter A to the value "FALSE", and submit, it works fine.
But then, if I change to the value "TRUE", and submit, the "FALSE" is still used.
If I change again to the value "TRUE", and I also set a value for the filter B, and submit, now the "TRUE" value is used.

I think that if all the values submitted are equal to the default value, the filter is not (or not correctly) submitted so no value are changed. Maybe we should be sure to always keep one field submitted (?).

cc @kirya-dev and @willemverspyck who works on this recently.

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because patch.

Closes #7250.

## Changelog

```markdown
### Fixed
- Filters are correctly updated if only default values are submitted.
```

### CODE REVIEW

**Comment by kirya-dev** [comment]
Hello!
Im dont thinking that this is good solution

**Comment by kirya-dev** [comment]
@VincentLanglet you can provide short screen cast with problem? (gif)

**Comment by VincentLanglet** [comment]
Default situation: 
image

I change a filter and submit
image

I remove the filter manually
image
Then if I submit it comes back
image

Because no filter are submitted so if you use filter persisters, you have this code 
```
                // if reset filters is asked, remove from storage
                if ('reset' === $this->getRequest()->query->get('filters')) {
                    $this->getFilterPersister()->reset($this->getCode());
                }

                // if no filters, fetch from storage
                // otherwise save to storage
                if (empty($filters)) {
                    $filters = $this->getFilterPersister()->get($this->getCode());
                } else {
                    $this->getFilterPersister()->set($this->getCode(), $filters);
                }
```
And i'm inside the `if (empty($filters)) {` case.

By simulating the reset, it solve the issue. I tried this code on my project.

**Comment by kirya-dev** [comment]
Im think problem in backend side

**Comment by kirya-dev** [comment]
You can provide url on each step?

**Comment by VincentLanglet** [comment]
First time I come
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?checkConsistency=
```
When I submit a new filter
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?filter%5BprocessedStatus%5D%5Bvalue%5D=operation.status.processed&checkConsistency=
```
When I remove the filter submitted (so I want to go back to the default values)
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?checkConsistency=
```
But the filter wasn't removed.

I don't think it's a backend issue. Because if after the url 
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?filter%5BprocessedStatus%5D%5Bvalue%5D=operation.status.processed&checkConsistency=
```
I would have go to another admin: 
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract/list
```
And then back to the initial admin
```
http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?checkConsistency=
```
The behavior is the one wanted, both filter are still filled, since I'm persisting the filter in the session.

The issue here is that if you submit no filter, the session is not updated.
There is nothing in the request saying that an empty submit has been done. So the admin cannot determine if we're loading the admin page from another page or not. That's why the only solution seems to add `filters=reset`

**Comment by kirya-dev** [comment]
Removed filter come back?

**Comment by VincentLanglet** [comment]
> Removed filter come back?

When I remove the value of a filter (and there is only the default value) and click on submit, yes it came back.
Because the url is `http://10.0.0.8/index.php/asv/admin/assurancevie/contract-operation-subscription/list?checkConsistency=`
And the persisterFilter use the filter of the session.

**Comment by kirya-dev** [comment]
How correctly change default filters via session?
It seems FilterPersister is extra feature...
Or we can redirect to correct link if has persisted values on link with persisted params ?

**Comment by VincentLanglet** [comment]
> How correctly change default filters via session?

IMHO, that's the correct way.
I want to submit the default value, either i add them in the url, either I set `filters=reset`.

So if you want to keep the feature which remove default value from the url, `filters=reset` is the solution.

There is a front hack to not fully submit the filter form, so it's a front bug, with a front fix.
The back end shouldn't be modified.

**Comment by kirya-dev** [comment]
Hmm. You solution can be exists)

Me seams will shorter change type of reset button to hidden to submit

File: src/Resources/public/Admin.js [hunk 1]
```diff
@@ -340,10 +340,20 @@ var Admin = {
...
+
+                if (JSON.stringify(defaults[field.name] || defaultValue) === JSON.stringify($(field).val() || defaultValue)) {
                     field.removeAttribute('name');
                 }
             });
+
+            // Simulate a reset if no value is different from the default ones.
+            if ($form.find('[name*=filter]').length === 0) {
+                var field = $(document.createElement("input"));
+                field.attr("name", 'filters').attr("type","hidden");
```
Comment (phansys) [inline src/Resources/public/Admin.js:353]: Just for consistency:
```suggestion
                field.attr('name', 'filters').attr('type', 'hidden');
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Resources/public/Admin.js (hunk 1 above)
Comment (VincentLanglet) [reply src/Resources/public/Admin.js]: Sure, done

**Review by phansys — APPROVED** [review]

**Comment by kirya-dev** [comment]
I have an idea, what else getDefaultFilterParameters must return merged params with persistFilter?

**Comment by VincentLanglet** [comment]
> I have an idea, what else getDefaultFilterParameters must return merged params with persistFilter?

I don't see how it solve the issue and I don't think it's a good idea.
This would say `getDefaultFilterParameters` will change with the request.
I expect a `getDefault` method to always return the same result.

**Comment by kirya-dev** [comment]
> This would say getDefaultFilterParameters will change with the request.

No. In my idea `getDefaultFilterParameters` must only merge. Dont reset or set.

**Comment by VincentLanglet** [comment]
> > This would say getDefaultFilterParameters will change with the request.
> 
> No. In my idea `getDefaultFilterParameters` must only merge. Dont reset or set.

This is already the case.

**Comment by kirya-dev** [comment]
> This is already the case
Can show me this?

**Comment by VincentLanglet** [comment]
> I see this in master

Yes, so it's only a merge.

> In my idea getDefaultFilterParameters must only merge

File: src/Resources/public/Admin.js [hunk 2]
```diff
@@ -340,10 +340,20 @@ var Admin = {
...
+                var defaultValue = field.multiple ? [] : '';
+
+                if (JSON.stringify(defaults[field.name] || defaultValue) === JSON.stringify($(field).val() || defaultValue)) {
                     field.removeAttribute('name');
                 }
             });
+
+            // Simulate a reset if no value is different from the default ones.
+            if ($form.find('[name*=filter]').length === 0) {
+                var field = $(document.createElement("input"));
```
Comment (phansys) [inline src/Resources/public/Admin.js:352]:
```suggestion
                var field = $(document.createElement('input'));
```

**Review by phansys — DISMISSED** [review]

File: src/Resources/public/Admin.js [hunk 3]
```diff
@@ -340,10 +340,20 @@ var Admin = {
...
+                var defaultValue = field.multiple ? [] : '';
+
+                if (JSON.stringify(defaults[field.name] || defaultValue) === JSON.stringify($(field).val() || defaultValue)) {
                     field.removeAttribute('name');
                 }
             });
+
+            // Simulate a reset if no value is different from the default ones.
+            if ($form.find('[name*=filter]').length === 0) {
+                var field = $(document.createElement('input'));
```
Comment (phansys) [inline src/Resources/public/Admin.js:352]:
```suggestion
                var field = jQuery(document.createElement('input'));
```
See #7262.

File: src/Resources/public/Admin.js [hunk 4]
```diff
@@ -340,10 +340,20 @@ var Admin = {
 
             // Keep only changed values
             $form.find('[name*=filter]').each(function (i, field) {
-                if (JSON.stringify(defaults[field.name] || '') === JSON.stringify($(field).val())) {
+                var defaultValue = field.multiple ? [] : '';
+
+                if (JSON.stringify(defaults[field.name] || defaultValue) === JSON.stringify($(field).val() || defaultValue)) {
```
Comment (phansys) [inline src/Resources/public/Admin.js:345]:
```suggestion
                if (JSON.stringify(defaults[field.name] || defaultValue) === JSON.stringify(jQuery(field).val() || defaultValue)) {
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Resources/public/Admin.js (hunk 3 above)
Comment (kirya-dev) [inline src/Resources/public/Admin.js:352]: All this method can be replace by 
```js
$form.append('<input type="filters" type="hidden" value="reset">');
```

Comment (VincentLanglet) [reply src/Resources/public/Admin.js]: Done

File: src/Resources/public/Admin.js (hunk 4 above)
Comment (VincentLanglet) [reply src/Resources/public/Admin.js]: Done

**Comment by jordisala1991** [comment]
Imo we should try to not introduce more jquery, can you provide a vanilla javascript implementation?

**Comment by VincentLanglet** [comment]
> Imo we should try to not introduce more jquery, can you provide a vanilla javascript implementation?

I'm not really introducing jquery here. I'm
- Fixing the `if (JSON.stringify(defaults[field.name] || '') === JSON.stringify(jQuery(field).val())) {` code.
- Adding call on an already existing variable `$form`.

Rewriting my code with vanilla JS would only make sens if I rewrite the whole function without Jquery.
It's a lot of extra-work for a bug-fix.
And since I don't work a lot with JS, especially vanilla, I would take me a lot of time and I cannot promise it will still work.
So I'd rather not to do this.

**Review by phansys — APPROVED** [review]

**Comment by kirya-dev** [comment]
Its can usefull to you
https://gist.github.com/joyrexus/7307312

**Comment by VincentLanglet** [comment]
Can we merge this @sonata-project/contributors ?

**Pull request merged**